### PR TITLE
[AC-5504] Initial Allocation Analyzer

### PR DESCRIPTION
**Pull request by jonkiparsky**
This PR adds a class which reads in a scenario (output from generate.py) and an allocation (output from allocate.py) and 

To test: 
run the allocate.py script on some scenario: `python3 allocate.py example.csv > tmp.out`
in the python shell
import the AllocationAnalyzer and use the `quick_setup` function to load the scenario and allocation:
```
>>> from classes.allocation_analyzer import *
>>> aa = quick_setup()
>>> counts = aa.analyze(aa.completed)
>>> summary = aa.summarize(counts)
>>> print ("\n".join(["%s: %s" %(key, val) for key, val in summary.items()]))
total_reads: 10390
Investor_reads: 1782
female_reads: 2458
Other_reads: 2099
...
```

Some changes to previous functionality:
- summary data removed from allocate.py output in order to allow csv reader to read it
- Event headers are now universal across the class
- pulled out the name munging for startups and judges

Things to do, now, later, or never:
- script this basic interaction
- metrics probably want to be objects
- tighter cohesion with Allocator and Generator (ability to talk to those objects directly rather than using csvs as intermediary)
- better analysis
- we're not dealing with judge niceness yet
- etc

UPDATES:
Added script. `python3 analyze.py` produces summaries of assigned and completed numbers to stdout, takes optional filenames for scenario and allocation csv files. 
Note: I'm assuming that these scripts are not intended as anything but developer shim, so I'm not making a whole lot of effort to make them nice

### CODE REVIEW

**Comment by mc-nathan** [comment]
Attempting to test I get:
```
>>> from classes.allocation_analyzer import *
>>> aa = quick_setup()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "classes/allocation_analyzer.py", line 124, in quick_setup
    aa.read_scenario_from_csv('example.csv')
  File "classes/allocation_analyzer.py", line 62, in read_scenario_from_csv
    self.process_scenario_from_csv(reader)
  File "classes/allocation_analyzer.py", line 72, in process_scenario_from_csv
    judge = Judge(data=row)
  File "classes/judge.py", line 13, in __init__
    super().__init__()
TypeError: super() takes at least 1 argument (0 given)
```

**Comment by mc-nathan** [comment]
Got the functional test to work.

File: classes/property.py [hunk 1]
```diff
@@ -10,7 +10,7 @@ def __init__(self, name, values=None):
         Property.all_properties.append(self)
 
 
-name = Property("name")
+name = Property("name",)
```
Comment (mc-nathan) [inline classes/property.py]: Remove comma or justify with a comment.

File: classes/judge.py [hunk 2]
```diff
@@ -46,3 +45,6 @@ def add_startup(self, startup):
     def mark_as_done(self):
         Event(action="done", subject=self)
         self.remaining = 0
+
+    def is_female(self):
```
Comment (mc-nathan) [inline classes/judge.py]: Not a blocker, but we should move away from special casing "female".

File: classes/entity.py [hunk 3]
```diff
@@ -52,6 +52,9 @@ def add_fields_to_name(self, fields):
     def _generated_name(self):
         return "%s-%s" % (self.type, self.properties['id'])
 
+    def __getitem__(self, key):
+        return self.properties.get(key, "BLANK")
```
Comment (mc-nathan) [inline classes/entity.py]: This feels like an odd default.  At minimum I would make it a constant and if it has to be not be None, then I would make it a little more distinct.  Maybe "<BLANK>".  "<NULL>" would be more natural to me than "<BLANK>", but that's a matter of personal preference.

File: classes/allocation_analyzer.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            summary['max %s' % metric] = val
+        for metric, val in list(mins.items()):
+            summary['min %s' % metric] = -val
+
+        summary['total_applications'] = total_applications
+        summary['total_judges'] = total_judges
+        return summary
+
+
+def quick_setup():
```
Comment (mc-nathan) [inline classes/allocation_analyzer.py]: Not a blocker, but I'd rather there was a separate script so we could do something like:
```
python3 analyze.py example.csv tmp.out
```

File: classes/allocation_analyzer.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+from classes.startup import Startup
+Assignment = namedtuple("Assignment", ["judge", "startup"])
+
+
+def increment_assignment_count(startup, metric_name, analysis):
+    name = startup['name']
+    analysis[name][metric_name] += 1
+
+
+def female_judge(assignment, read_counts):
```
Comment (mc-nathan) [inline classes/allocation_analyzer.py]: Again, I'd rather this wasn't special cased.

File: classes/allocation_analyzer.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                maxes[metric] = max(maxes[metric], count)
+                mins[metric] = max(mins[metric], -count)
+        total_applications = len(self.startups)
+        total_judges = len(self.judges)
+        for metric, count in list(summary.items()):
+            summary['average %s' % metric] = count / total_applications
+        for metric, val in list(maxes.items()):
+            summary['max %s' % metric] = val
+        for metric, val in list(mins.items()):
+            summary['min %s' % metric] = -val
```
Comment (mc-nathan) [inline classes/allocation_analyzer.py]: I'd like to see counts especially for the min values so we know how much we are missing by.

File: classes/allocation_analyzer.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def program_match(assignment, read_counts):
+    judge, startup = assignment
+    metric_name = "home_program_reads"
+    if startup['program'] == judge['program']:
+        increment_assignment_count(startup, metric_name, read_counts)
+
+
+def industry_match(assignment, read_counts):
```
Comment (mc-nathan) [inline classes/allocation_analyzer.py]: The parallels with program_match strongly suggests that this should be refactored.  Ideally, there should be no feature specific code, but it should all be inferred from the features and values that get used in the input files.

**Review by mc-nathan — CHANGES_REQUESTED** [review]
Overall works as described, but there's some valuable refactoring that should get done that is also pointed to by CodeClimate.

**Comment by mc-nathan** [comment]
I generally agree with the suggested improvements.  Personally I would not prioritize getting away from the CSV usage, but that's because I find simple shell commands easier to deal with than figuring out everything in the python shell.  Otherwise I would handle them in the order described.

I would also add some count information at least to the mins and there would be no harm in adding them to the maxs.  I think that would allow us to actually understand if we are meeting the criteria or by how much we are missing them.

**Comment by jonkiparsky** [comment]
@mc-nathan I've rejiggered to have each metric report on how many applications missed the specified target, and removed the mins since they seemed like a poor way to get at this information

File: classes/judge.py (hunk 2 above)
Comment (jonkiparsky) [reply classes/judge.py]: done

File: classes/allocation_analyzer.py (hunk 6 above)
Comment (jonkiparsky) [reply classes/allocation_analyzer.py]: See below, replaced min counts with "missed"

**Comment by mc-nathan** [comment]
```
$ make code-check
analyze.py:19:1: E303 too many blank lines (3)
analyze.py:28:1: W391 blank line at end of file
classes/allocation_analyzer.py:17:1: E302 expected 2 blank lines, found 1
classes/allocation_analyzer.py:25:61: E231 missing whitespace after ','
classes/allocation_analyzer.py:48:45: W291 trailing whitespace
classes/allocation_analyzer.py:79:29: E221 multiple spaces before operator
classes/allocation_analyzer.py:80:80: E501 line too long (85 > 79 characters)
classes/gender_distribution_metric.py:3:1: E302 expected 2 blank lines, found 1
classes/gender_distribution_metric.py:13:1: W293 blank line contains whitespace
classes/gender_distribution_metric.py:13:1: W391 blank line at end of file
classes/industry_match_metric.py:3:1: E302 expected 2 blank lines, found 1
classes/industry_match_metric.py:8:35: W291 trailing whitespace
classes/judge_role_distribution_metric.py:3:1: E302 expected 2 blank lines, found 1
classes/judge_role_distribution_metric.py:6:1: E302 expected 2 blank lines, found 1
classes/judge_role_distribution_metric.py:14:1: W293 blank line contains whitespace
classes/judge_role_distribution_metric.py:17:1: W293 blank line contains whitespace
classes/judge_role_distribution_metric.py:17:1: W391 blank line at end of file
classes/metric.py:9:5: E303 too many blank lines (2)
classes/metric.py:11:1: W293 blank line contains whitespace
classes/metric.py:27:1: W293 blank line contains whitespace
classes/metric.py:29:80: E501 line too long (81 > 79 characters)
classes/metric.py:35:1: W293 blank line contains whitespace
classes/metric.py:35:1: W391 blank line at end of file
classes/program_match_metric.py:3:1: E302 expected 2 blank lines, found 1
classes/program_match_metric.py:4:19: E211 whitespace before '('
classes/program_match_metric.py:6:1: W293 blank line contains whitespace
classes/total_reads_metric.py:3:1: E302 expected 2 blank lines, found 1
classes/total_reads_metric.py:6:1: W391 blank line at end of file
make: *** [code-check] Error 1
```

**Review by mc-nathan — APPROVED** [review]

**Pull request merged**